### PR TITLE
step 4: Bridging the deployment of the gnosis safe via omni bridge

### PR DESCRIPTION
**Pull request by josojo**
Bridging the deployment of the gnosis safe via omni bridge

### Test Plan
unit tests only

### CODE REVIEW

**Comment by josojo** [comment]
Sorry, I meant to open it only as a draft. No need to review it, although it is close to completion

File: src/ts/proposal.ts [hunk 1]
```diff
@@ -210,6 +225,44 @@ async function generateBridgeTokenToGnosisChainTx(
...
+    "IAMB",
+    bridgeSettings.arbitraryMessageBridgeETH,
+  );
+  const deploySafeDeterministicOnGnosisChain = {
+    to: ambForeign.address,
+    value: "0",
+    data: ambForeign.interface.encodeFunctionData("requireToPassMessage", [
+      to,
+      data,
+      2000000, // Max value is 5M, 2M should be sufficient for gnosis safe deployment.
```
Comment (fedgiac) [inline src/ts/proposal.ts:259]:
```suggestion
      2000000, // Max value is currently 2.5M, 2M should be sufficient for gnosis safe deployment.
```

File: test/proposal.test.ts [hunk 2]
```diff
@@ -54,6 +60,12 @@ describe("proposal", function () {
     multiTokenMediatorETH = await OmniBridgeTransferSimulator.connect(
       deployer,
     ).deploy();
+    const IAMB = await artifacts.readArtifact("IAMB");
+    arbitraryMessageBridge = await waffle.deployMockContract(
+      deployer,
+      IAMB.abi,
+    );
+    await arbitraryMessageBridge.mock.requireToPassMessage.returns(messageID);
```
Comment (fedgiac) [inline test/proposal.test.ts:68]: I think there should be a test that checks the input of `requireToPassMessage` somewhere, even if it copies some of the code use to generate the bytecode. I see that the input is manually decoded later, so maybe we could move the mock there and use the decoded values.

File: test/proposal.test.ts [hunk 3]
```diff
@@ -306,3 +320,130 @@ describe("proposal", function () {
...
+        },
+        settings.cowDao,
+        gnosisSafeDefaults,
+        hre.ethers,
+      );
+      expect(
+        await hre.ethers.provider.getCode(expectedCowDaoAddress),
+      ).to.be.equal("0x");
+      const cutoffData = 266;
+      const cutoffTo = 34;
```
Comment (fedgiac) [inline test/proposal.test.ts:422]: You might be able to do something like:
```
const signatureBytes = 4;
const [to, data, gas] = defaultAbiCoder.decode(["address", "bytes", "uint256"], arrayify(bridgedGnosisSafeDeployment.data).slice(signatureBytes))
```

File: test/proposal.test.ts [hunk 4]
```diff
@@ -306,3 +320,130 @@ describe("proposal", function () {
...
+          "0x" +
+          bridgedGnosisSafeDeployment.data.substring(cutoffTo, 40 + cutoffTo),
+        data:
+          "0x" +
+          bridgedGnosisSafeDeployment.data.substring(
+            cutoffData,
+            1288 + cutoffData,
+          ),
+        gasPrice: 545019933,
+        gasLimit: 3008448,
```
Comment (fedgiac) [inline test/proposal.test.ts:435]: Is this needed?

File: test/proposal.test.ts [hunk 5]
```diff
@@ -306,3 +320,130 @@ describe("proposal", function () {
...
+            arbitraryMessageBridgeETH:
+              settings.bridge.arbitraryMessageBridgeETH,
+          },
+          settings.cowDao,
+          gnosisSafeDefaults,
+          hre.ethers,
+        ),
+      ).to.be.rejectedWith(Error);
+    });
+    it("has the txData that allows to deploy the safe with correct threshold and owners", async function () {
```
Comment (fedgiac) [inline test/proposal.test.ts:408]: I don't understand how the numbers were computed, but otherwise it's a good test!

File: test/safe.ts [hunk 6]
```diff
@@ -40,6 +42,35 @@ export class GnosisSafeManager {
     );
   }
 
+  public static async initDeterministic(deployer: Signer): Promise<GnosisSafeManager> {
```
Comment (fedgiac) [inline test/safe.ts:45]: Unfortunately the state is not cleared every time we start a new unrelated test, so calling this function more than once will be an issue in the future. This is something that can't be tackled now, just for you to know that you can't rely on this in the future without some extra checks that the deployment destination address won't exist.

**Review by fedgiac — COMMENTED** [review]
Logic looks sound. :heavy_check_mark:

File: test/proposal.test.ts (hunk 3 above)
Comment (fedgiac) [reply test/proposal.test.ts:424]: I'll do this in a separate commit.

File: test/proposal.test.ts (hunk 4 above)
Comment (josojo) [reply test/proposal.test.ts]: Yes, though I am not sure why

File: test/proposal.test.ts (hunk 2 above)
Comment (josojo) [reply test/proposal.test.ts:70]: Yes, this was a good point! I am not sure about best practice, but now we are testing:
- (1) the AMB will be called with certain parameters
- (2) thse parameters from (1) are executable and generate the correct safe

I think with that we are safe now

File: test/proposal.test.ts [hunk 7]
```diff
@@ -306,3 +323,167 @@ describe("proposal", function () {
...
+        ]),
+      ).to.be.reverted;
+      await arbitraryMessageBridge.mock.requireToPassMessage
+        .withArgs(to, data, gasLimit)
+        .returns(messageID);
+      await expect(
+        execSafeTransaction(gnosisDao, bridgedGnosisSafeDeployment, [
+          gnosisDaoOwner,
+        ]),
+      ).to.not.be.reverted;
```
Comment (fedgiac) [inline test/proposal.test.ts:428]: Nice! This resolves my [comment from here](https://github.com/gnosis/cow-token/pull/72/files#r792486292).

**Review by fedgiac — APPROVED** [review]
Looks good!

**Pull request merged**